sabnzbd/dirscanner.py

```diff
@@ -115,99 +109,169 @@ def save(self):
+        if not self.loop:
+            return
+
+        if not (dirscan_dir := self.dirscan_dir):
+            return
+
+        asyncio.run_coroutine_threadsafe(self.scan_async(dirscan_dir), self.loop)
+
+    async def scan_async(self, dirscan_dir: str):
+        """Do one scan of the watched folder"""
+        async with self.lock:
+            if sabnzbd.PAUSED_ALL:
+                return
+
+            files: Set[str] = set()
+            futures: Set[asyncio.Task] = set()
+
+            for path, catdir, stat_tuple in self.get_suspected_files(dirscan_dir):
+                files.add(path)
+
+                if path in self.ignored or path in self.suspected:
+                    continue
+
+                if stat_tuple.st_size > 0:
+                    futures.add(asyncio.create_task(self.when_stable_add_nzbfile(path, catdir, stat_tuple)))
+                    await asyncio.sleep(0)
 
             # Remove files from the bookkeeping that are no longer on the disk
-            clean_file_list(self.ignored, folder, files)
-            clean_file_list(self.suspected, folder, files)
-
-        if not self.busy:
-            self.busy = True
-            dirscan_dir = self.dirscan_dir
-            if dirscan_dir and not sabnzbd.PAUSED_ALL:
-                run_dir(dirscan_dir, None)
-
-                try:
-                    dirscan_list = os.listdir(dirscan_dir)
-                except OSError:
-                    if not self.error_reported:
-                        logging.error(T("Cannot read Watched Folder %s"), filesystem.clip_path(dirscan_dir))
-                        self.error_reported = True
-                    dirscan_list = []
-
-                cats = config.get_categories()
-                for dd in dirscan_list:
-                    dpath = os.path.join(dirscan_dir, dd)
-                    if os.path.isdir(dpath) and dd.lower() in cats:
-                        run_dir(dpath, dd.lower())
-            self.busy = False
+            # Wait for the paths found in this scan to finish
+            await asyncio.gather(clean_file_list(self.ignored, files), clean_file_list(self.suspected, files), *futures)
+
+    async def scanner(self):
+        """Periodically scan the directory and add NZB files to the queue"""
+        while True:
+            if not (dirscan_speed := self.dirscan_speed):
+                break
+
+            if not (dirscan_dir := self.dirscan_dir):
+                break
+
+            try:
+                await self.scan_async(dirscan_dir)
+            except:
+                pass
+
+            await asyncio.sleep(dirscan_speed)
+
+    async def shutdown(self):
+        """Cancel all tasks and stop the loop"""
+        loop = asyncio.get_event_loop()
+
+        # Get all tasks except for this one
+        tasks = filter(lambda task: task is not asyncio.current_task(), asyncio.all_tasks())
+
+        # Cancel them all
+        [task.cancel() for task in tasks]
```

Comment:
I used filter instead of list comprehension, it's still a oneliner but I think easier to understand?

The second could be a for loop if you'd prefer, but that one seems simpler?

Reply:
Second one seems better as a loop.